### DISCUSSION TITLE
Help setting up secured cluster with Terraform

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#14989] [Q&A] [answered]
Hey guys

I'm helping a client setup their Raven cluster with Terraform but documentation is a bit thin.

1. As far as I can tell, the only way to setup a secured cluster is to create a configuration ZIP file with a local instance and delegate the ZIP file? It's not possible to just define the certificate, node names and ports, yes?

2. I presume that in the secured scenario `http_port` will actually be the HTTPS port, right?

3. What is the recommended way to create a client cert in a TF deployment?

4. If the HTTP port is publicly available for access to Studio, does the TCP port also need to be publicly available or is it used only for inter-node comms?

Thanks!

**Comment by ayende** [comment, marked as answer]
1/ You can define the things manually by providing the `settings.json` and `pfx`, yes. You can also generate the zip file using `rvn create-setup-package` command to do that without the studio.

2/ Yes

3/ Use `rvn create-setup-package` and then apply that during the Terraform run. Once that happens, you can create a client certificate using the _server_ certificate

4/ The TCP is used for inter cluster communication as well as for other reasons. Such as subscriptions, external replication, etc.

**Reply by georgiosd** [reply]
Thank you @ayende! When you say to create a client cert from the server cert, is there another command for that?

**Reply by ayende** [reply]
You can do that using `PutClientCertificateOperation`, using the server certificate.

**Reply by georgiosd** [reply]
Curious - why does `create-setup-package` require the license when it's not actually included in the package and then required by Terraform too?

It would also be great to provide an example of the setup.json file somewhere - I had to read the source code to replicate it.

**Reply by ayende** [reply]
The `create-setup-package` generates a certificate for a domain that RavenDB owns. We need to avoid giving the same domain to multiple people. We use the license to associate a domain, to ensure that there won't be any situation where someone can "hijack" your domain.

**Reply by georgiosd** [reply]
Makes sense! In that case, it would be great to make it optional for `own-certificate`

**Reply by ayende** [reply]
Pretty sure it is, no?

**Reply by georgiosd** [reply]
It's not :-\

setup.json:
```
{
  ...
 "license": null,
 ....
}
```
Then:
```
PS C:\Users\georgiosd\Desktop> C:\RavenDB\Server\rvn.exe create-setup-package -m own-certificate -o test.zip -s setup.json -c cert.pfx

Error: License must be set
```

**Reply by ayende** [reply]
https://issues.hibernatingrhinos.com/issue/RavenDB-19394/rvn-create-setup-package-shouldnt-require-a-license-for-own-certificate
